src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -0,0 +1,584 @@
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_delete_blob_request(
+  bucket_name const& name, object_key const& key) {
+    // DELETE /{container-id}/{blob-id} HTTP/1.1
+    // Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    const auto target = fmt::format("/{}/{}", name(), key().string());
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::delete_);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+    header.insert("x-ms-delete-snapshots", "include");
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_list_blobs_request(
+  const bucket_name& name,
+  std::optional<object_key> prefix,
+  [[maybe_unused]] std::optional<object_key> start_after,
+  std::optional<size_t> max_keys) {
+    // GET /{container-id}?restype=container&comp=list&prefix={prefix}...
+    // ...&max_results{max_keys}
+    // HTTP/1.1 Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    auto target = fmt::format("/{}?restype=container&comp=list", name());
+    if (prefix) {
+        target += fmt::format("&prefix={}", prefix.value()().string());
+    }
+
+    if (max_keys) {
+        target += fmt::format("&max_results={}", max_keys.value());
+    }
+
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::get);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+abs_client::abs_client(
+  const abs_configuration& conf,
+  ss::lw_shared_ptr<const cloud_roles::apply_credentials> apply_credentials)
+  : _requestor(conf, std::move(apply_credentials))
+  , _client(conf)
+  , _probe(conf._probe) {}
+
+abs_client::abs_client(
+  const abs_configuration& conf,
+  const ss::abort_source& as,
+  ss::lw_shared_ptr<const cloud_roles::apply_credentials> apply_credentials)
+  : _requestor(conf, std::move(apply_credentials))
+  , _client(conf, &as, conf._probe, conf.max_idle_time)
+  , _probe(conf._probe) {}
+
+ss::future<> abs_client::stop() { return _client.stop(); }
+
+void abs_client::shutdown() { _client.shutdown(); }
+
+template<typename T>
+ss::future<result<T, error_outcome>> abs_client::send_request(
+  ss::future<T> request_future,
+  const bucket_name& bucket,
+  const object_key& key) {
```

Comment:
prefer to pass values to coroutines

Reply:
Yeah. This is my biggest gripe with the interface for the S3 client. If we changed to pass-by-value for these two strings, we'd end up making quite a few copies (we can't always move these args in).

As far as I understand this is safe in two scenarios. I'll outline them below, but please keep me honest here:
1. The client functions get called from a chain of `co_await(s)` and all the args used in the chain are local to the callsite, have a longer lifetime than the duration of the whole chain, or are created in the call itself. This is the case in the existing non-test code. `cloud_storage::remote` is the only caller of the clients (outside of the test code) and all calls to it are `co_await(ed)`.

For example:
```
constexpr bucket_name = "bucket";

foo() {
auto path = gen_path(...);
co_await _remote.upload_segment(
  bucket_name,
  path,
  created_here{...},
  ...);
}

remote::upload_segment(const auto& bucket, const auto& key, ...) {
  co_await abs_client.put_object(bucket, key, ...); 
}
```
2. In our test code, we make calls on the remote without using `co_await`. For example, the code below will end up calling into the client. 
```
remote.delete_object(bucket, object_key(path), ...).get();
```

I think what keeps this safe is a combination const-lifetime extension to keep the temporaries alive until a co_await is reached, seastar threads, and seastar's chaining of continuations. I've not analysed these as well, so I'm a bit fuzzy here.

Reply:
you are correct about these cases. the nice thing about coroutines, i think, is that if that's all that is used, then everything is pretty safe (although cpp core guideline does say to only pass values to coroutines). in redpanda, however, we have a mixture of coroutines and continuation passing style. the problem then becomes: all of the call sites need to be examined to tell if a coroutine is used safely, because mixing those styles is error prone.

> If we changed to pass-by-value for these two strings, we'd end up making quite a few copies

what does "quite a few copies" mean? a client that is ostensibly making encrypted http network requests, it seems that copying some strings would be totally lost in the noise.

---

anyway, nothing to do here in this PR related to this topic, but something for us to consider. like coroutine lambdas, i'd expect that if we start to see lifetime problems that aren't caught in review, that we'll think about getting rid of the pattern.